src/i2c/src/i2c.rs

```diff
@@ -149,22 +161,29 @@ pub struct I2cSmbusIoctlData {
     data: *mut I2cSmbusData,
 }
 
-impl I2cSmbusIoctlData {
+pub struct SmbusMsg {
+    read_write: u8,
+    command: u8,
+    size: u32,
+    data: Option<I2cSmbusData>,
+}
+
+impl SmbusMsg {
     /// Based on Linux's drivers/i2c/i2c-core-smbus.c:i2c_smbus_xfer_emulated().
     ///
     /// These smbus related functions try to reverse what Linux does, only
     /// support basic modes (up to word transfer).
-    fn new(reqs: &mut [I2cReq]) -> Result<I2cSmbusIoctlData> {
+    fn new(reqs: &mut [I2cReq]) -> Result<SmbusMsg> {
         let mut data = I2cSmbusData {
             block: [0; I2C_SMBUS_BLOCK_MAX + 2],
         };
         let read_write: u8;
-        let size: u32;
+        let mut smbus_data = None;
 
         // Write messages have only one request message, while read messages
         // will have two (except for few special cases of I2C_SMBUS_QUICK and
         // I2C_SMBUS_BYTE, where only one request messages is sent).
-        match reqs.len() {
+        let size = match reqs.len() {
             // Write requests (with some exceptions as mentioned above)
```

Comment:
this seems a bit weird. Why have multiple bits of data and state rather than just returning Some(SmbusMsg) directly from processing the message?

Reply:
I am not sure I understood this comment well, the routine in question `SmbusMsg::new()` does return SmbusMsg only, we are just trying to find out what all the fields should look like. Can you explain a bit on how you think this can be improved ?